File: python/minecraft_bot/protocol/v763/packets/play/clientbound/multi_block_change.py
"""Packet `multi_block_change` (play/clientbound, id 0x43).

Updates many blocks within a single 16x16x16 chunk section in one
packet. ``chunk_coordinates`` is a 64-bit packed bitfield::

    [22 bits signed x | 22 bits signed z | 20 bits signed y]

Each ``records`` entry is a packed VarLong (encoded here as a varint
stream of long-style values, but minecraft-data calls them varints in
the schema; the in-protocol units are 22+8 = 30 bit packed
``(block_state_id << 12) | (rel_pos & 0xFFF)``).
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

from minecraft_bot.codec import Reader, Writer, varint

PACKET_ID = 0x43


def _sign_extend(value: int, bits: int) -> int:
    mask = (1 << bits) - 1
    value &= mask
    if value & (1 << (bits - 1)):
        value -= 1 << bits
    return value


@dataclass(frozen=True, slots=True)
class MultiBlockChange:
    chunk_section_x: int   # 22-bit signed
    chunk_section_z: int   # 22-bit signed
    chunk_section_y: int   # 20-bit signed
    records: tuple[int, ...]   # packed (block_state << 12) | rel_xyz


def decode(reader: Reader) -> MultiBlockChange:
    packed, = struct.unpack(">q", reader.read(8))
    cy = _sign_extend(packed & 0xFFFFF, 20)
    cz = _sign_extend((packed >> 20) & 0x3FFFFF, 22)
    cx = _sign_extend((packed >> 42) & 0x3FFFFF, 22)
    n = varint.read(reader)
    recs = tuple(varint.read(reader) for _ in range(n))
    return MultiBlockChange(
        chunk_section_x=cx, chunk_section_z=cz, chunk_section_y=cy, records=recs,
    )


def encode(packet: MultiBlockChange, writer: Writer) -> None:
    cx = packet.chunk_section_x & 0x3FFFFF
    cz = packet.chunk_section_z & 0x3FFFFF
    cy = packet.chunk_section_y & 0xFFFFF
    packed = (cx << 42) | (cz << 20) | cy
    if packed & (1 << 63):
        packed -= 1 << 64
    writer.write(struct.pack(">q", packed))
    varint.write(len(packet.records), writer)
    for r in packet.records:
        varint.write(r, writer)

File: rust/src/protocol/v763/packets/play/clientbound/multi_block_change.rs
//! Packet `multi_block_change` (play/clientbound). Hand-written.

use crate::codec::uuid_codec::{self as uuid_c, Uuid};
use crate::codec::{
    bitset, chat_component, identifier, nbt, position, slot, string_codec as string, varint,
    varlong, BytesReader, BytesWriter, Reader, Writer,
};
use crate::errors::ProtocolError;
use crate::protocol::v763::states::ConnectionState;
use crate::protocol::v763::ClientboundPacket;

pub const PACKET_ID: i32 = 0x43;

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct MultiBlockChange {
    pub chunk_section_x: i32,
    pub chunk_section_z: i32,
    pub chunk_section_y: i32,
    pub records: Vec<i64>,
}

fn sign_extend(v: i64, bits: i32) -> i32 {
    let mask: i64 = (1i64 << bits) - 1;
    let mut v = v & mask;
    if v & (1i64 << (bits - 1)) != 0 {
        v -= 1i64 << bits;
    }
    v as i32
}

impl MultiBlockChange {
    pub fn decode(reader: &mut BytesReader<'_>) -> Result<Self, ProtocolError> {
        let b = reader.read_exact(8)?;
        let packed = i64::from_be_bytes([b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7]]);
        let cy = sign_extend(packed & 0xFFFFF, 20);
        let cz = sign_extend((packed >> 20) & 0x3FFFFF, 22);
        let cx = sign_extend((packed >> 42) & 0x3FFFFF, 22);
        let n = varint::read(reader)? as usize;
        let mut records = Vec::with_capacity(n);
        for _ in 0..n {
            records.push(varlong::read(reader)?);
        }
        Ok(Self {
            chunk_section_x: cx,
            chunk_section_z: cz,
            chunk_section_y: cy,
            records,
        })
    }
}

impl ClientboundPacket for MultiBlockChange {
    fn state(&self) -> ConnectionState {
        ConnectionState::Play
    }
    fn packet_id(&self) -> i32 {
        PACKET_ID
    }
    fn encode(&self, writer: &mut BytesWriter) -> Result<(), ProtocolError> {
        let packed: i64 = ((self.chunk_section_x as i64 & 0x3FFFFF) << 42)
            | ((self.chunk_section_z as i64 & 0x3FFFFF) << 20)
            | (self.chunk_section_y as i64 & 0xFFFFF);
        writer.write_all(&packed.to_be_bytes())?;
        varint::write(self.records.len() as i32, writer)?;
        for r in &self.records {
            varlong::write(*r, writer)?;
        }
        Ok(())
    }
}
